horovod/run/gloo_run.py

```diff
@@ -234,79 +184,132 @@ def _exec_command(command, alloc_info, event):
     if settings.output_filename:
         _mkdir_p(settings.output_filename)
 
+    # start global rendezvous server and get port that it is listening on
+    rendezvous = RendezvousServer(settings.verbose)
+
+    # allocate processes into slots
+    hosts = parse_hosts(settings.hosts)
+    host_alloc_plan = get_host_assignments(hosts, settings.num_proc)
+
+    # start global rendezvous server and get port that it is listening on
+    global_rendezv_port = rendezvous.start_server()
+    rendezvous.httpd.init(host_alloc_plan)
+    run_command = get_run_command(command, server_ip, nics, global_rendezv_port)
+
+    slot_info_to_command = _slot_info_to_command_fn(run_command, env)
+    event = register_shutdown_event()
+    args_list = [[slot_info_to_command(slot_info), slot_info, [event]]
+                 for slot_info in host_alloc_plan]
+
     # If an error occurs in one thread, entire process will be terminated.
     # Otherwise, threads will keep running.
-    res = threads.execute_function_multithreaded(exec_command, args_list, block_until_all_done=True)
+    res = threads.execute_function_multithreaded(exec_command,
+                                                 args_list,
+                                                 block_until_all_done=True)
 
     for name, value in sorted(res.items(), key=lambda item: item[1][1]):
         exit_code, timestamp = value
         if exit_code != 0:
-            raise RuntimeError('Gloo job detected that one or more processes exited with non-zero '
+            raise RuntimeError('Horovod detected that one or more processes exited with non-zero '
                                'status, thus causing the job to be terminated. The first process '
                                'to do so was:\nProcess name: {name}\nExit code: {code}\n'
                                .format(name=name, code=exit_code))
 
 
-def gloo_run(settings, remote_host_names, nics, env, server_ip, command):
+def _get_min_start_hosts(settings):
+    # This function exists for the purpose of mocking in tests
+    return 2 if settings.max_np > settings.num_proc and not settings.nics else 1
+
+
+def gloo_run(settings, nics, env, server_ip, command):
     # Each thread will use ssh command to launch the job on each remote host. If an
     # error occurs in one thread, entire process will be terminated. Otherwise,
     # threads will keep running and ssh session.
-    exec_command = _exec_command_fn(settings, remote_host_names)
+    exec_command = _exec_command_fn(settings)
     launch_gloo(command, exec_command, settings, nics, env, server_ip)
+
+
+def gloo_run_elastic(settings, env, command):
+    # Make the output directory if it does not exist
+    if settings.output_filename:
+        _mkdir_p(settings.output_filename)
+
+    rendezvous = RendezvousServer(settings.verbose)
+    driver = ElasticDriver(rendezvous, settings.discovery,
+                           settings.min_np, settings.max_np,
```

Comment:
There is no way to set `start_timeout` here, defaults to 10 minutes, `settings` should have a way to wire this number through to this place. The existing `settings.timeout` is a `timeout.Timeout` instance.

Reply:
`ElasticDriver` could reuse `settings.timeout` as this measures time since the start of the horovod cluster.

Reply:
@tgaddair I have addressed this with commit d842e2a1a29200c372f67a263156d2599226eb78, please review

Reply:
Looks good, thanks!